### PR TITLE
Implement verbosity feature for kubeadm

### LINKED ISSUE #340: Implement different verbosity levels in kubeadm
From the upgrades proposal and other design discussions it is clear that it's time for kubeadm to implement a `--v` flag now that kubeadm implements more advanced features where more logging would be useful.

This would be a great contribution for someone that want to get their hands dirty with something...

### PR DESCRIPTION
**Pull request by vbmade2000**
[WIP] Adds verbosity feature to init command hierarchy of kubeadm utility.

**What this PR does / why we need it**:
Implements verbosity feature to kubeadm

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#340

**Special notes for your reviewer**:
I will be splitting this work into a smaller PR to keep it separate and clean.

**Release note**:

```release-note
Implements verbosity logging feature for kubeadm commands
```

### CODE REVIEW

**Review by luxas — COMMENTED** [review]
cc @kubernetes/sig-cluster-lifecycle-pr-reviews @ncdc @thockin @sttts @kubernetes/sig-cli-pr-reviews 

What kind of logging solution would you prefer for kubeadm to implement?

**Comment by sttts** [comment]
> What kind of logging solution would you prefer for kubeadm to implement?

glog is still our logging library. Structured logging is mentioned once in a while, also at KubeCon again. But I haven't seen further investment into that.

**Comment by timothysc** [comment]
I like logrus and we use it on our tools.

**Comment by vbmade2000** [comment]
@timothysc Should We change it to use logrus then ? I am waiting for some conclusion so that I can work further on this issue.

**Comment by dixudx** [comment]
> Should We change it to use logrus then ?

@vbmade2000 Not now. Currently we still use `glog` in Kubernetes.

**Comment by vbmade2000** [comment]
@dixudx Ok. So can I proceed further with glog ?

**Comment by dixudx** [comment]
> Ok. So can I proceed further with glog ?

@sttts Any objection on still using `glog`?

**Comment by jamiehannaford** [comment]
Since logrus is already [listed as a dep](https://github.com/kubernetes/kubernetes/blob/cc22b10278a3e63cdbdcbd8522856a666af1e2ed/vendor/BUILD#L329), and the eventual aim is for kubeadm to live in its own repo, is there any reason why not to use it if most of us deem it a better tool? Just wondering if adhering to precedent is worth it here.

**Comment by vbmade2000** [comment]
@luxas @timothysc Any conclusion ? This issue is pending since long time.

**Review by timothysc — CHANGES_REQUESTED** [review]
I think it's fine to stay with glog. I'm not a fan of the explicit leveling for most of the default statements, they should all be Infof for now and we can decide later what their verbosity is.

**Comment by vbmade2000** [comment]
@timothysc Sure.

> I'm not a fan of the explicit leveling for most of the default statements, they should all be Infof for now.
Didn't get it.

File: cmd/kubeadm/app/cmd/init.go [hunk 1]
```diff
@@ -120,6 +121,7 @@ func NewCmdInit(out io.Writer) *cobra.Command {
 		Short: "Run this command in order to set up the Kubernetes master.",
 		Run: func(cmd *cobra.Command, args []string) {
 			var err error
+
```
Comment (jamiehannaford) [inline cmd/kubeadm/app/cmd/init.go]: nit: can you remove these extra lines? doesn't seem related to changeset. prefer to keep diffs small :)

File: cmd/kubeadm/app/cmd/init.go [hunk 2]
```diff
@@ -224,12 +226,14 @@ func AddInitOtherFlags(flagSet *flag.FlagSet, cfgPath *string, skipPreFlight, sk
...
 		`Specify the CRI socket to connect to.`,
 	)
+
 }
 
 // NewInit validates given arguments and instantiates Init struct with provided information.
 func NewInit(cfgPath string, cfg *kubeadmapi.MasterConfiguration, ignorePreflightErrors sets.String, skipTokenPrint, dryRun bool, criSocket string) (*Init, error) {
 
 	if cfgPath != "" {
+		glog.V(1).Infof("[init] Reading config file from : " + cfgPath)
```
Comment (jamiehannaford) [inline cmd/kubeadm/app/cmd/init.go]: So we only have two verbosity levels? 0 (default) and 1, right? might be good to document this somewhere

File: cmd/kubeadm/app/cmd/init.go [hunk 3]
```diff
@@ -301,29 +309,34 @@ func (i *Init) Run(out io.Writer) error {
...
-		fmt.Println("[externalca] The file 'ca.key' was not found, yet all other certificates are present. Using external CA mode - certificates or kubeconfig will not be generated.")
+		glog.Infoln("[externalca] The file 'ca.key' was not found, yet all other certificates are present. Using external CA mode - certificates or kubeconfig will not be generated.")
 	}
 
 	// Temporarily set cfg.CertificatesDir to the "real value" when writing controlplane manifests
 	// This is needed for writing the right kind of manifests
 	i.cfg.CertificatesDir = realCertsDir
 
 	// PHASE 3: Bootstrap the control plane
+	glog.V(1).Infof("[init] Bootstraping the control plane")
```
Comment (jamiehannaford) [inline cmd/kubeadm/app/cmd/init.go]: bootstrapping

File: cmd/kubeadm/app/cmd/init.go (hunk 2 above)
Comment (jamiehannaford) [reply cmd/kubeadm/app/cmd/init.go]: Also, should be: `from:`

File: cmd/kubeadm/app/cmd/init.go [hunk 4]
```diff
@@ -240,31 +244,34 @@ func NewInit(cfgPath string, cfg *kubeadmapi.MasterConfiguration, ignorePrefligh
 	}
 
 	// Set defaults dynamically that the API group defaulting can't (by fetching information from the internet, looking up network interfaces, etc.)
+	glog.V(1).Infof("[init] Setting dynamic defaults")
 	err := configutil.SetInitDynamicDefaults(cfg)
 	if err != nil {
 		return nil, err
 	}
 
+	glog.V(1).Infof("[init] Validating kubernetes version")
```
Comment (jamiehannaford) [inline cmd/kubeadm/app/cmd/init.go]: Kubernetes (capitalised)

File: cmd/kubeadm/app/cmd/init.go [hunk 5]
```diff
@@ -240,31 +244,34 @@ func NewInit(cfgPath string, cfg *kubeadmapi.MasterConfiguration, ignorePrefligh
...
-	fmt.Printf("[init] Using Kubernetes version: %s\n", cfg.KubernetesVersion)
-	fmt.Printf("[init] Using Authorization modes: %v\n", cfg.AuthorizationModes)
+	glog.Infof("[init] Using Kubernetes version: %s\n", cfg.KubernetesVersion)
+	glog.Infof("[init] Using Authorization modes: %v\n", cfg.AuthorizationModes)
 
 	// Warn about the limitations with the current cloudprovider solution.
 	if cfg.CloudProvider != "" {
-		fmt.Println("[init] WARNING: For cloudprovider integrations to work --cloud-provider must be set for all kubelets in the cluster.")
-		fmt.Println("\t(/etc/systemd/system/kubelet.service.d/10-kubeadm.conf should be edited for this purpose)")
+		glog.Warningln("[init] For cloudprovider integrations to work --cloud-provider must be set for all kubelets in the cluster.")
```
Comment (jamiehannaford) [inline cmd/kubeadm/app/cmd/init.go]: Here you're terminating with a period `.`, but in other places you remove them. Can we pick one style and make that consistent?

File: cmd/kubeadm/app/cmd/init.go [hunk 6]
```diff
@@ -378,60 +396,71 @@ func (i *Init) Run(out io.Writer) error {
...
 	}
 
 	// Create/update RBAC rules that makes the nodes to rotate certificates and get their CSRs approved automatically
+	glog.V(1).Infof("[init] Creating/Updating RBAC rules for rotating certificate")
 	if err := nodebootstraptokenphase.AutoApproveNodeCertificateRotation(client); err != nil {
 		return err
 	}
 
 	// Create the cluster-info ConfigMap with the associated RBAC rules
+	glog.V(1).Infof("[init] Creating bootstrap config map")
```
Comment (jamiehannaford) [inline cmd/kubeadm/app/cmd/init.go]: nit: configmap

File: cmd/kubeadm/app/cmd/init.go [hunk 7]
```diff
@@ -378,60 +396,71 @@ func (i *Init) Run(out io.Writer) error {
...
 	if err := clusterinfophase.CreateBootstrapConfigMapIfNotExists(client, adminKubeConfigPath); err != nil {
 		return fmt.Errorf("error creating bootstrap configmap: %v", err)
 	}
+	glog.V(1).Infof("[init] Creating ClsuterInfo RBAC rules")
```
Comment (jamiehannaford) [inline cmd/kubeadm/app/cmd/init.go]: ClusterInfo

**Review by jamiehannaford — CHANGES_REQUESTED** [review]
Thanks for submitting this! Generally LGTM, but I think we should document flag usage, especially the different verbosity levels

File: cmd/kubeadm/app/cmd/init.go (hunk 1 above)
Comment (vbmade2000) [reply cmd/kubeadm/app/cmd/init.go]: Sure. This is not a final push. This is just to show approach and get confirmation from members but thanks, I'll keep that in mind  :)

File: cmd/kubeadm/app/cmd/init.go (hunk 2 above)
Comment (vbmade2000) [reply cmd/kubeadm/app/cmd/init.go]: Good observation for `from:`. I'll take care of that from now on.

File: cmd/kubeadm/app/cmd/init.go (hunk 4 above)
Comment (vbmade2000) [reply cmd/kubeadm/app/cmd/init.go]: Got it.

File: cmd/kubeadm/app/cmd/init.go (hunk 5 above)
Comment (vbmade2000) [reply cmd/kubeadm/app/cmd/init.go]: Sure.

File: cmd/kubeadm/app/cmd/init.go (hunk 3 above)
Comment (vbmade2000) [reply cmd/kubeadm/app/cmd/init.go]: Got it.

File: cmd/kubeadm/app/cmd/init.go (hunk 6 above)
Comment (vbmade2000) [reply cmd/kubeadm/app/cmd/init.go]: Got it.

File: cmd/kubeadm/app/cmd/init.go (hunk 7 above)
Comment (vbmade2000) [reply cmd/kubeadm/app/cmd/init.go]: Got it.

**Review by timothysc — APPROVED** [review]
/approve

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/57661)

**Review by timothysc — APPROVED** [review]
Thank you for the patch and your patience. 

/lgtm 
/approve
